src/hotspot/share/gc/shenandoah/heuristics/shenandoahReactiveHeuristics.cpp

```diff
@@ -0,0 +1,308 @@
+                         z_score,
+                         byte_size_in_proper_unit(_available.avg()), proper_unit_for_byte_size(_available.avg()),
+                         byte_size_in_proper_unit(_available.sd()), proper_unit_for_byte_size(_available.sd()));
+  }
+
+  // In the case when a concurrent GC cycle completes successfully but with an
+  // unusually small amount of available memory we will adjust our trigger
+  // parameters so that they are more likely to initiate a new cycle.
+  // Conversely, when a GC cycle results in an above average amount of available
+  // memory, we will adjust the trigger parameters to be less likely to initiate
+  // a GC cycle.
+  //
+  // The z-score we've computed is in no way statistically related to the
+  // trigger parameters, but it has the nice property that worse z-scores for
+  // available memory indicate making larger adjustments to the trigger
+  // parameters. It also results in fewer adjustments as the application
+  // stabilizes.
+  //
+  // In order to avoid making endless and likely unnecessary adjustments to the
+  // trigger parameters, the change in available memory (with respect to the
+  // average) at the end of a cycle must be beyond these threshold values.
+  if (z_score < LOWEST_EXPECTED_AVAILABLE_AT_END ||
+      z_score > HIGHEST_EXPECTED_AVAILABLE_AT_END) {
+    // The sign is flipped because a negative z-score indicates that the
+    // available memory at the end of the cycle is below average. Positive
+    // adjustments make the triggers more sensitive (i.e., more likely to fire).
+    // The z-score also gives us a measure of just how far below normal. This
+    // property allows us to adjust the trigger parameters proportionally.
+    //
+    // The `100` here is used to attenuate the size of our adjustments. This
+    // number was chosen empirically. It also means the adjustments at the end of
+    // a concurrent cycle are an order of magnitude smaller than the adjustments
+    // made for a degenerated or full GC cycle (which themselves were also
+    // chosen empirically).
+    adjust_last_trigger_parameters(z_score / -100);
+  }
+}
+
+void ShenandoahReactiveHeuristics::record_success_degenerated() {
+  ShenandoahHeuristics::record_success_degenerated();
+  // Adjust both trigger's parameters in the case of a degenerated GC because
+  // either of them should have triggered earlier to avoid this case.
+  adjust_margin_of_error(DEGENERATE_PENALTY_SD);
+  adjust_spike_threshold(DEGENERATE_PENALTY_SD);
+}
+
+void ShenandoahReactiveHeuristics::record_success_full() {
+  ShenandoahHeuristics::record_success_full();
+  // Adjust both trigger's parameters in the case of a full GC because
+  // either of them should have triggered earlier to avoid this case.
+  adjust_margin_of_error(FULL_PENALTY_SD);
+  adjust_spike_threshold(FULL_PENALTY_SD);
+}
+
+static double safe_adjust(double value, double adjustment, double min, double max) {
```

Comment:
Seems like this is just:

```
static double saturate(double value, double min, double max) {
  return MAX2(MIN2(value, max), min);
}
```

...in fact, we probably use this pattern without the explicit method everywhere else.

Reply:
Refactored this to the pattern you suggested.